### PR TITLE
Implement service-level payload

### PR DESCRIPTION
**Pull request by mytang0**
## What is the purpose of the change
Implement service-level payload.  Issue #10773  

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.

- [x] Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [x] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java [hunk 1]
```diff
@@ -31,11 +31,7 @@
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicBoolean;
 
-import static org.apache.dubbo.common.constants.CommonConstants.APPLICATION_KEY;
-import static org.apache.dubbo.common.constants.CommonConstants.GROUP_KEY;
-import static org.apache.dubbo.common.constants.CommonConstants.INTERFACE_KEY;
-import static org.apache.dubbo.common.constants.CommonConstants.TIMEOUT_KEY;
-import static org.apache.dubbo.common.constants.CommonConstants.VERSION_KEY;
+import static org.apache.dubbo.common.constants.CommonConstants.*;
```
Comment (AlbumenJ) [inline dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java:34]: Do not use comma import.
You can import ide code style config in `codestyle` folder.

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/HeaderExchangeChannel.java [hunk 2]
```diff
@@ -126,11 +126,16 @@ public CompletableFuture<Object> request(Object request, int timeout, ExecutorSe
...
-        req.setData(request);
+        Request req;
+        if (request instanceof Request) {
+            req = (Request) request;
+        } else {
+            // create request.
+            req = new Request();
+            req.setVersion(Version.getProtocolVersion());
+            req.setTwoWay(true);
+            req.setData(request);
```
Comment (AlbumenJ) [inline dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/HeaderExchangeChannel.java:137]: What is the purpose of changing this?

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java (hunk 1 above)
Comment (mytang0) [reply dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java:34]: Ok, this problem has been fixed.

Comment (AlbumenJ) [reply dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java:34]: dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/HeaderExchangeChannel.java (hunk 2 above)
Comment (mytang0) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/HeaderExchangeChannel.java:137]: > What is the purpose of changing this?

In order to get the payload on the provider side, I added an attribute to the Request. The value of payload will be set in DubboInvoker.


The differences are as follows:
image

File: dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java (hunk 1 above)
Comment (mytang0) [reply dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java:34]: > MetadataInfoTest

Sorry, I'll fix it right away.

File: dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/HeaderExchangeChannel.java (hunk 2 above)
Comment (mytang0) [reply dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/exchange/support/header/HeaderExchangeChannel.java:137]: Done.

File: dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java (hunk 1 above)
Comment (mytang0) [reply dubbo-metadata/dubbo-metadata-api/src/test/java/org/apache/dubbo/metadata/MetadataInfoTest.java]: Done, submitted.

File: dubbo-common/src/main/java/org/apache/dubbo/config/annotation/DubboService.java [hunk 3]
```diff
@@ -323,4 +323,9 @@
      * @return
      */
     String executor() default "";
+
+    /**
+     * Payload max length.
+     */
+    String payload() default "";
```
Comment (wxbty) [inline dubbo-common/src/main/java/org/apache/dubbo/config/annotation/DubboService.java:330]: Why not use the int type, consistent with the previous one?

Comment (mytang0) [reply dubbo-common/src/main/java/org/apache/dubbo/config/annotation/DubboService.java:330]: > Why not use the int type, consistent with the previous one?

Thank you. Yes, I started using int and later made changes to be consistent with dubbo.xsd

**Review by wxbty — CHANGES_REQUESTED** [review]
Adding a payload to the request at the communication layer is unnecessary and risky. The verification of the payload is in the following fragment. The payload itself is obtained through the url. We may modify the corresponding value in the current url through filter or other interception.

```java
protected static int getPayload(Channel channel) {
    if (channel != null && channel.getUrl() != null) {
        return channel.getUrl().getParameter(Constants.PAYLOAD_KEY, Constants.DEFAULT_PAYLOAD);
    }
    return Constants.DEFAULT_PAYLOAD;
}
```

**Comment by mytang0** [comment]
> Adding a payload to the request at the communication layer is unnecessary and risky. The verification of the payload is in the following fragment. The payload itself is obtained through the url. We may modify the corresponding value in the current url through filter or other interception.
> 
> ```java
> protected static int getPayload(Channel channel) {
>     if (channel != null && channel.getUrl() != null) {
>         return channel.getUrl().getParameter(Constants.PAYLOAD_KEY, Constants.DEFAULT_PAYLOAD);
>     }
>     return Constants.DEFAULT_PAYLOAD;
> }
> ```

I am thinking about the following:
1. Channel is shared, so modifying it is complicated and risky
2. Channel's URL payload is a protocol payload
3. The payload property of Request is not passed over the network , so it will not increase network overhead and affect the packaging and unpacking of messages

**Comment by wxbty** [comment]
> 

channel.getUrl() The attributes here come from app-config, service-config and other external configurations, so at least they are shared at the service level.
Since the parameters are obtained from the url, maybe we can achieve the payload configuration at the service level without doing anything. As the following example:

```java
@DubboService(parameters = {"payload", "3000"})
public class GameServiceImpl2 implements GameService {
    public String play(String name) {
        return "hello 2181:" + name;
    }
}
```


image

**Comment by wxbty** [comment]
The configuration of parameters parameters may not be obvious to users, so it should be necessary to add attributes to @DubboService

**Comment by mytang0** [comment]
Thank you very much for your suggestion. It is true that the solution you suggested is the best implementation when the channel is not shared. Consumers share channels through provider address (host + port), each channel holds the url of the service that created it. Therefore, in the sharing mode, it cannot be guaranteed that the payload obtained through the channel is accurate.

**Comment by wxbty** [comment]
```java
/**
 * Get dubbo channel by netty channel through channel cache.
 * Put netty channel into it if dubbo channel don't exist in the cache.
 *
 * @param ch      netty channel
 * @param url
 * @param handler dubbo handler that contain netty's handler
 * @return
 */
static NettyChannel getOrAddChannel(Channel ch, URL url, ChannelHandler handler) {
    if (ch == null) {
        return null;
    }
    NettyChannel ret = CHANNEL_MAP.get(ch);
    if (ret == null) {
        NettyChannel nettyChannel = new NettyChannel(ch, url, handler);
        if (ch.isActive()) {
            nettyChannel.markActive(true);
            ret = CHANNEL_MAP.putIfAbsent(ch, nettyChannel);
        }
        if (ret == null) {
            ret = nettyChannel;
        }
    } else {
        ret.markActive(true);
    }
    return ret;
}
```

@AlbumenJ    NettyChannel seems to have a bug here. The key of CHANNEL_MAP is indeed the original channel of netty, and the url is not included, which means that when calling service2, the channel(created by first service) obtained contains the url of service1. The same goes for actual testing. Maybe we can solve it like this: the key of CHANNEL_MAP uses Channel+url instead of Channel

**Comment by mytang0** [comment]
`private List<ReferenceCountExchangeClient> getSharedClient(URL url, int connectNum) {
        String key = url.getAddress();

        Object clients = referenceClientMap.get(key);
        if (clients instanceof List) {
            List<ReferenceCountExchangeClient> typedClients = (List<ReferenceCountExchangeClient>) clients;
            if (checkClientCanUse(typedClients)) {
                batchClientRefIncr(typedClients);
                return typedClients;
            }
        }
......
}
`

In this place, cause the url may already be inaccurate.  I think Url is used as a property container, and the remoting layer mainly transmits the configuration properties of the remoting layer.  Therefore, when dealing with sharing, Url seems to be less important.

**Comment by AlbumenJ** [comment]
> ```java
> /**
>  * Get dubbo channel by netty channel through channel cache.
>  * Put netty channel into it if dubbo channel don't exist in the cache.
>  *
>  * @param ch      netty channel
>  * @param url
>  * @param handler dubbo handler that contain netty's handler
>  * @return
>  */
> static NettyChannel getOrAddChannel(Channel ch, URL url, ChannelHandler handler) {
>     if (ch == null) {
>         return null;
>     }
>     NettyChannel ret = CHANNEL_MAP.get(ch);
>     if (ret == null) {
>         NettyChannel nettyChannel = new NettyChannel(ch, url, handler);
>         if (ch.isActive()) {
>             nettyChannel.markActive(true);
>             ret = CHANNEL_MAP.putIfAbsent(ch, nettyChannel);
>         }
>         if (ret == null) {
>             ret = nettyChannel;
>         }
>     } else {
>         ret.markActive(true);
>     }
>     return ret;
> }
> ```
> 
> @AlbumenJ NettyChannel seems to have a bug here. The key of CHANNEL_MAP is indeed the original channel of netty, and the url is not included, which means that when calling service2, the channel(created by first service) obtained contains the url of service1. The same goes for actual testing. Maybe we can solve it like this: the key of CHANNEL_MAP uses Channel+url instead of Channel

We should prevent to get parameters from channel url. Channel should be shared among services.

File: dubbo-demo/dubbo-demo-xml/dubbo-demo-xml-provider/src/main/resources/spring/dubbo-provider.xml [hunk 4]
```diff
@@ -44,5 +44,8 @@
                    ref="restDemoService" protocol="rest"/>
     <dubbo:service delay="5000" version="1.0.0" timeout="5000" interface="org.apache.dubbo.demo.TripleService"
                    ref="tripleService" protocol="tri"/>
+    <dubbo:service delay="5000" version="2.0.0" group="greeting" timeout="5000"
+                   interface="org.apache.dubbo.demo.GreetingService"
+                   ref="greetingService" protocol="dubbo" payload="1024"/>
```
Comment (CrazyHZM) [inline dubbo-demo/dubbo-demo-xml/dubbo-demo-xml-provider/src/main/resources/spring/dubbo-provider.xml:49]: the same as above.

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

**Comment by CrazyHZM** [comment]
@mytang0 Regarding the payload configuration of this service level, is it possible to bind different services to different protocol configs to achieve the purpose?

File: dubbo-demo/dubbo-demo-annotation/dubbo-demo-annotation-provider/src/main/java/org/apache/dubbo/demo/provider/DemoServiceImpl.java [hunk 5]
```diff
@@ -25,7 +25,7 @@
 
 import java.util.concurrent.CompletableFuture;
 
-@DubboService
+@DubboService(payload = "1024")
```
[the remark this answers is not in the document]
Comment (mytang0) [reply dubbo-demo/dubbo-demo-annotation/dubbo-demo-annotation-provider/src/main/java/org/apache/dubbo/demo/provider/DemoServiceImpl.java:28]: Ok, i fix

**Comment by mytang0** [comment]
> @mytang0 Regarding the payload configuration of this service level, is it possible to bind different services to different protocol configs to achieve the purpose?

Yes, it can be achieved by adding channels, but there is a little violation of channel sharing.

**Comment by mytang0** [comment]
PTAL @chickenlj

**Review by chickenlj — APPROVED** [review]
LGTM.

**Review by AlbumenJ — APPROVED** [review]

**Pull request merged**